tests/semantic_analyser.cpp

```diff
@@ -8,237 +11,150 @@
-{
-  auto bpftrace = get_mock_bpftrace();
-  return test_for_warning(*bpftrace, input, warning, invert, safe_mode);
-}
-
-ast::ASTContext test(BPFtrace &bpftrace,
-                     bool mock_has_features,
-                     const std::string &input,
-                     int expected_result,
-                     std::string_view expected_error = "",
-                     bool safe_mode = true,
-                     bool has_child = false)
-{
-  std::string local_input(input);
-  if (!local_input.empty() && local_input[0] == '\n')
-    local_input = local_input.substr(1); // Remove initial '\n'
-  ast::ASTContext ast("stdin", local_input);
-
-  std::stringstream msg;
-  msg << "\nInput:\n" << input << "\n\nOutput:\n";
-
-  bpftrace.safe_mode_ = safe_mode;
-  bpftrace.feature_ = std::make_unique<MockBPFfeature>(mock_has_features);
-  if (has_child) {
-    bpftrace.cmd_ = "not-empty"; // Used by SemanticAnalyser.
+struct Options {
+  bool safe_mode = true;
+  bool with_child = false;
+  bool features = true;
+};
+struct Mock {
+  BPFtrace &bpftrace;
+};
+struct Warning {
+  std::string_view str;
+};
+struct NoWarning {
+  std::string_view str;
+};
+struct Error {
+  std::string_view str;
+};
+struct ExpectedAST {
+  std::string_view str;
+};
+
+template <typename T, typename First, typename... Ts>
+std::optional<T> extract(First &&arg, Ts &&...rest)
+{
+  if constexpr (std::is_same_v<std::decay_t<First>, T>) {
+    return arg;
   }
-
-  // N.B. See above.
-  auto ok = ast::PassManager()
-                .put(ast)
-                .put(bpftrace)
-                .add(CreateParsePass())
-                .add(ast::CreateMacroExpansionPass())
-                .add(ast::CreateParseAttachpointsPass())
-                .add(ast::CreateFieldAnalyserPass())
-                .add(ast::CreateClangParsePass())
-                .add(ast::CreateCMacroExpansionPass())
-                .add(ast::CreateFoldLiteralsPass())
-                .add(ast::CreateMapSugarPass())
-                .add(ast::CreateSemanticPass())
-                .run();
-
-  // Reproduce the full string.
-  std::stringstream out;
-  ast.diagnostics().emit(out);
-
-  if (expected_result == 0) {
-    // Accept no errors.
-    EXPECT_TRUE(ok && ast.diagnostics().ok()) << msg.str() << out.str();
-  } else {
-    // Accept any failure result.
-    EXPECT_FALSE(ok && ast.diagnostics().ok()) << msg.str();
-  }
-  if (expected_error.data() && !expected_error.empty()) {
-    if (!expected_error.empty() && expected_error[0] == '\n')
-      expected_error.remove_prefix(1); // Remove initial '\n'
-    EXPECT_EQ(out.str(), expected_error);
+  if constexpr (sizeof...(Ts) != 0) {
+    return extract<T, Ts...>(std::forward<Ts>(rest)...);
   }
-
-  return ast;
+  return std::nullopt;
 }
 
-ast::ASTContext test(BPFtrace &bpftrace,
-                     const std::string &input,
-                     bool safe_mode = true)
+template <typename T>
+std::optional<T> extract()
```

Comment:
Why do we need this extract that accepts no params? Is it the case when `...args` passed to `test` doesn't exist that this is called when invoking extract?

Reply:
Correct.